gno.land/pkg/gnoweb/markdown/ext_links.go

```diff
@@ -0,0 +1,191 @@
+	path   string
+}
+
+// linkTypeInfo contains information about a link type
+type linkTypeInfo struct {
+	tooltip string
+	icon    string
+	class   string
+}
+
+// attr represents an HTML attribute
+type attr struct {
+	name  string
+	value string
+}
+
+// RegisterFuncs registers the renderer functions
+func (r *linkRenderer) RegisterFuncs(reg renderer.NodeRendererFuncRegisterer) {
+	reg.Register(ast.KindLink, r.renderLink)
+}
+
+// isSamePackage checks if the link points to the same package
+// - Link is an anchor (starts with #)
+// - Link is relative (no leading / and no protocol -> root link)
+// - Link points to the same package in /r/ path
+// - Link points to the same package in /p/ path
+// - Link points to the same package in /u/ path
+func isSamePackage(dest, pathWithoutR string) bool {
+	return strings.HasPrefix(dest, "#") ||
+		(!strings.HasPrefix(dest, "/") && !strings.Contains(dest, "://")) ||
+		strings.Contains(dest, "/r/"+pathWithoutR) ||
+		strings.Contains(dest, "/p/"+pathWithoutR) ||
+		strings.Contains(dest, "/u/"+pathWithoutR)
+}
+
+// Error messages for invalid link formats
+var (
+	ErrLinkInvalidURL = errors.New("invalid URL format")
+)
+
+// writeHTMLAttr writes an HTML attribute
+func writeHTMLAttr(w util.BufWriter, name, value string) {
+	_, _ = w.WriteString(" ")
+	_, _ = w.WriteString(name)
+	_, _ = w.WriteString(`="`)
+	_, _ = w.WriteString(value)
+	_, _ = w.WriteString(`"`)
+}
+
+// writeHTMLTag writes an HTML tag with its attributes
+func writeHTMLTag(w util.BufWriter, tag string, attrs []attr) {
+	_, _ = w.WriteString("<" + tag)
+	for _, a := range attrs {
+		writeHTMLAttr(w, a.name, a.value)
+	}
+	_, _ = w.WriteString(">")
+}
+
+// detectLinkType detects the type of link based on the destination
+func detectLinkType(dest, domain, path string) (string, bool, error) {
+	// Extract the package name from the path (e.g., "r/test/foo" -> "test")
+	pathWithoutR := strings.TrimPrefix(path, "r/")
+	if idx := strings.Index(pathWithoutR, "/"); idx != -1 {
+		pathWithoutR = pathWithoutR[:idx]
+	}
+
+	// Check if the link is external:
+	// - Contains a protocol (e.g., http://, https://)
+	// - Contains a domain different from the current one
+	if strings.Contains(dest, "://") && !strings.Contains(dest, "://"+domain) {
+		return "external", false, nil
+	}
+
+	// Check if the link is a package link
+	if isSamePackage(dest, pathWithoutR) {
+		return "package", strings.Contains(dest, "$help"), nil
+	}
+
+	// All other links are internal
+	return "internal", strings.Contains(dest, "$help"), nil
+}
+
+var linkTypes = map[string]linkTypeInfo{
+	"external": {tooltipExternalLink, iconExternalLink, classLinkExternal},
+	"internal": {tooltipInternalLink, iconInternalLink, classLinkInternal},
+	"tx":       {tooltipTxLink, iconTxLink, classLinkTx},
+}
+
+// renderLink renders a link node
+func (r *linkRenderer) renderLink(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	n := node.(*ast.Link)
```

Comment:
This cast is dangerous, as it can be nil. you should check with `n, ok := node.(*ast.Link)` I don't even understand why it's not crashing yet. this is weird.